# Compare externally sources

And compare against OneMap, with OneMap seeming to be the best...

In [154]:
%load_ext kedro.ipython
%reload_kedro --env=test
%load_ext autoreload
%autoreload 2
%config IPCompleter.use_jedi=False
from IPython.core.interactiveshell import InteractiveShell
import os
import pandas as pd
InteractiveShell.ast_node_interactivity = "all"
os.chdir(context.project_path)
catalog = context.catalog
params = context.params

The kedro.ipython extension is already loaded. To reload it, use:
  %reload_ext kedro.ipython


[01/22/25 22:21:26] INFO     Resolved project path as:                                              ]8;id=807660;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py\__init__.py]8;;\:]8;id=609599;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py#178\178]8;;\
                             /Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geoc                
                             oding.                                                                                
                             To set a different path, run '%reload_kedro <project_root>'                           

                    INFO     Kedro project singapore-postcode-geocoding                             ]8;id=723700;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py\__init__.py]8;;\:]8;id=940848;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py#144\144]8;;\

                    INFO     Defined global variable 'context', 'session', 'catalog' and            ]8;id=290582;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py\__init__.py]8;;\:]8;id=746770;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py#145\145]8;;\
                             'pipelines'                                                                           

                    INFO     Registered line magic 'run_viz'                                        ]8;id=415716;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py\__init__.py]8;;\:]8;id=185100;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/ipython/__init__.py#151\151]8;;\

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


## Inspect OpenData

In [91]:
postal_codes = catalog.load("open_data_postal_code")
postal_codes = postal_codes.loc[postal_codes["country_code"] == "SG"]
postal_codes

[01/22/25 21:21:32] INFO     Loading data from open_data_postal_code (ParquetDataset)...        ]8;id=79098;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=68429;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py#389\389]8;;\

country_code postal_code       place_name admin_name1 admin_code1  \
0                SG      806485     Sunrise Lane        None        None   
1                SG      806526    Sunrise Drive        None        None   
2                SG      806582    Sunrise Drive        None        None   
3                SG      806605    Sunrise Close        None        None   
4                SG      806608    Sunrise Close        None        None   
...             ...         ...              ...         ...         ...   
122212           SG      138681       Dover Rise        None        None   
122213           SG      138698      Woking Road        None        None   
122214           SG      138811  Whitchurch Road        None        None   
122215           SG      138883    Weyhill Close        None        None   
122216           SG      138941  Westbourne Road        None        None   

       admin_name2 admin_code2 admin_name3 admin_code3  latitude  longitude  \
0             None        None        None        None    1.3898   103.8546   
1             None        None        None        None    1.3880   103.8572   
2             None        None        None        None    1.3905   103.8549   
3             None        None        None        None    1.3875   103.8576   
4             None        None        None        None    1.3871   103.8570   
...            ...         ...         ...         ...       ...        ...   
122212        None        None        None        None    1.3056   103.7838   
122213        None        None        None        None    1.2967   103.7952   
122214        None        None        None        None    1.2938   103.7982   
122215        None        None        None        None    1.2986   103.7930   
122216        None        None        None        None    1.2943   103.7965   

        accuracy                                        coordinates  
0            NaN  b'\x01\x01\x00\x00\x00X\xca2\xc4\xb1\xf6Y@[\xb...  
1            NaN  b'\x01\x01\x00\x00\x00f\x88c]\xdc\xf6Y@h\x91\x...  
2            NaN  b'\x01\x01\x00\x00\x00mV}\xae\xb6\xf6Y@sh\x91\...  
3            NaN  b'\x01\x01\x00\x00\x00-C\x1c\xeb\xe2\xf6Y@3333...  
4            NaN  b'\x01\x01\x00\x00\x00\x02+\x87\x16\xd9\xf6Y@o...  
...          ...                                                ...  
122212       NaN  b'\x01\x01\x00\x00\x00\xb5\xa6y\xc7)\xf2Y@\x88...  
122213       NaN  b'\x01\x01\x00\x00\x00\xdeq\x8a\x8e\xe4\xf2Y@:...  
122214       NaN  b'\x01\x01\x00\x00\x00\xb3\xeas\xb5\x15\xf3Y@l...  
122215       NaN  b'\x01\x01\x00\x00\x00\x98n\x12\x83\xc0\xf2Y@\...  
122216       NaN  b'\x01\x01\x00\x00\x00\xe5\xd0"\xdb\xf9\xf2Y@\...  

[121154 rows x 13 columns]

Looked at two and the addresses and postal codes seem to coincide, geographically...

## Postcode base via getdata

It has fairly complete address info, concatenated, unfortunately.

In [5]:
postcode_address = catalog.load("sg_postcode_based_via_getdata")

[01/22/25 19:12:57] INFO     Loading data from sg_postcode_based_via_getdata (CSVDataset)...    ]8;id=108110;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=148955;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py#389\389]8;;\

## Merge data

There are some duplicates here. Based on name spellings, but they are all in very close proximity.

In [6]:
postal_codes.loc[postal_codes["postal_code"].duplicated(keep=False)].sort_values(
    "postal_code"
)

country_code postal_code             place_name admin_name1  \
85772            SG      098634      St. John'S Island        None   
22671            SG      098634    Saint John'S Island        None   
64924            SG      098659    Saint John'S Island        None   
30554            SG      098659      St. John'S Island        None   
64947            SG      117555        Science Drive 4        None   
43839            SG      117555  Lower Kent Ridge Road        None   
3685             SG      178959    Saint Andrew'S Road        None   
85975            SG      178959      St. Andrew'S Road        None   
120620           SG      187954          Waterloo Link        None   
3693             SG      187954        Waterloo Street        None   
79480            SG      257996     St. Martin'S Drive        None   
42219            SG      257996   Saint Martin'S Drive        None   
26733            SG      322004      St. George'S Lane        None   
90753            SG      322004    Saint George'S Lane        None   
90755            SG      327919     Saint Wilfred Road        None   
95615            SG      327919       St. Wilfred Road        None   
55851            SG      327920       St. Wilfred Road        None   
59495            SG      327920     Saint Wilfred Road        None   
7928             SG      327976     St. Michael'S Road        None   
86290            SG      327976   Saint Michael'S Road        None   
25727            SG      328002     St. Michael'S Road        None   
82108            SG      328002   Saint Michael'S Road        None   
107711           SG      328069       St. Francis Road        None   
34116            SG      328069     Saint Francis Road        None   
52795            SG      328070     Saint Francis Road        None   
99525            SG      328070       St. Francis Road        None   
104517           SG      363202         Woodleigh Link        None   
38975            SG      363202    Bidadari Park Drive        None   
56024            SG      368051      Mount Vernon Road        None   
73374            SG      368051           Bartley Road        None   
35245            SG      529538     Tampines Street 32        None   
54415            SG      529538     Tampines Central 1        None   
31634            SG      556083      Serangoon Central        None   
55198            SG      556083     Serangoon Avenue 2        None   
10748            SG      574117      Bright Hill Drive        None   
67685            SG      574117       Bright Hill Road        None   
98994            SG      719023      Neo Tiew Crescent        None   
4771             SG      719023          Neo Tiew Road        None   

       admin_code1 admin_name2 admin_code2 admin_name3 admin_code3  latitude  \
85772         None        None        None        None        None    1.2172   
22671         None        None        None        None        None    1.2172   
64924         None        None        None        None        None    1.2162   
30554         None        None        None        None        None    1.2161   
64947         None        None        None        None        None    1.2960   
43839         None        None        None        None        None    1.2962   
3685          None        None        None        None        None    1.2923   
85975         None        None        None        None        None    1.2924   
120620        None        None        None        None        None    1.2983   
3693          None        None        None        None        None    1.2983   
79480         None        None        None        None        None    1.3056   
42219         None        None        None        None        None    1.3056   
26733         None        None        None        None        None    1.3221   
90753         None        None        None        None        None    1.3222   
90755         None        None        None        None        None    1.3246

In [7]:
postal_codes_merge = postal_codes.assign(
    postal_code_merge_key=postal_codes["postal_code"].astype("string").str.zfill(6)
)
postcode_address_merge = postcode_address.assign(
    postal_code_merge_key=postcode_address["postcode"].astype("string").str.zfill(6)
)
postal_merge = postal_codes_merge.merge(
    postcode_address_merge, on="postal_code_merge_key", how="outer", validate="m:1"
)

In [8]:
postal_merge["postcode"].isna().sum() / postal_merge.shape[0]

np.float64(0.07532081086107495)

In [9]:
postal_merge["postal_code"].isna().sum() / postal_merge.shape[0]

np.float64(0.09872419564813092)

In [10]:
(
    ~(postal_merge["postal_code"].isna() | postal_merge["postcode"].isna())
).sum() / postal_merge.shape[0]

np.float64(0.8259549934907942)

In [11]:
(~(postal_merge["postal_code"].isna() | postal_merge["postcode"].isna())).sum()

np.int64(111029)

In [12]:
postal_merge_match = postal_merge.loc[
    ~(postal_merge["postal_code"].isna() | postal_merge["postcode"].isna())
]
results_view = postal_merge_match[["postal_code", "place_name", "address"]]
results_view

postal_code                  place_name  \
20          018955             Bayfront Avenue   
21          018956             Bayfront Avenue   
25          018960                 Marina View   
28          018965           Central Boulevard   
32          018971             Bayfront Avenue   
...            ...                         ...   
134402      829752  Ponggol Seventeenth Avenue   
134403      829753  Ponggol Seventeenth Avenue   
134404      829754  Ponggol Seventeenth Avenue   
134405      829755  Ponggol Seventeenth Avenue   
134406      829756  Ponggol Seventeenth Avenue   

                                                  address  
20      Bayfront Avenue, 8, Singapore, Bayfront, Marin...  
21      Bayfront Avenue, 10, Singapore, Bayfront, Mari...  
25      Marina View, 8, Singapore, Marina, Marina Sout...  
28      Marina Bay Suites, Central Boulevard, 3, Singa...  
32      Bayfront Avenue, 1, Singapore, Bayfront, Marin...  
...                                                   ...  
134402  Ponggol Seventeenth Avenue, 221F, Singapore, S...  
134403  Ponggol Seventeenth Avenue, 221G, Singapore, S...  
134404  Ponggol Seventeenth Avenue, 221H, Singapore, S...  
134405  Ponggol Seventeenth Avenue, 221J, Singapore, S...  
134406  Ponggol Seventeenth Avenue, 221K, Singapore, S...  

[111029 rows x 3 columns]

The above gives us 111029 addresses, well short of the estimated 140k addresses.

## Compare against 2020 onemap scrape

In [13]:
onemap = catalog.load("one_map_scrape")
onemap_merge = onemap.assign(
    postal_code_merge_key=onemap["POSTAL"].astype("string").str.zfill(6)
)
onemap_merge

[01/22/25 19:13:13] INFO     Loading data from one_map_scrape (CSVDataset)...                   ]8;id=593130;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=409580;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py#389\389]8;;\

ADDRESS   BLK_NO  \
0        1 STRAITS BOULEVARD SINGAPORE CHINESE CULTURA...        1   
1        11A STRAITS BOULEVARD TEMPORARY SITE OFFICE S...      11A   
2               5A MARINA GARDENS DRIVE SINGAPORE 018910        5A   
3        2 CENTRAL BOULEVARD CENTRAL BOULEVARD TOWERS ...        2   
4        21 PARK STREET MARINA BAY MRT STATION SINGAPO...       21   
...                                                   ...      ...   
141843       1A UPPER CHANGI ROAD NORTH SINGAPORE 886129        1A   
141844                 100A KRANJI LOOP SINGAPORE 887327      100A   
141845             A PASIR PANJANG ROAD SINGAPORE 887328         A   
141846   GATE C7 AIRPORT CARGO ROAD CHANGI ANIMAL AND ...  GATE C7   
141847   AIRPORT BOULEVARD CHANGI AIRPORT TERMINAL 2 B...      NaN   

                                         BUILDING  LATITUDE   LONGITUDE  \
0               SINGAPORE CHINESE CULTURAL CENTRE  1.275805  103.849615   
1                           TEMPORARY SITE OFFICE  1.274950  103.851665   
2                                             NIL  1.279587  103.868956   
3                        CENTRAL BOULEVARD TOWERS  1.279744  103.851591   
4                   MARINA BAY MRT STATION (NS27)  1.276409  103.854595   
...                                           ...       ...         ...   
141843                                        NIL  1.352316  103.966549   
141844                                        NIL  1.433629  103.758648   
141845                                        NIL  1.277170  103.795840   
141846  CHANGI ANIMAL AND PLANT QUARANTINE CENTRE  1.375315  103.996683   
141847         CHANGI AIRPORT TERMINAL 2 BUILDING  1.354969  103.989067   

        POSTAL                ROAD_NAME            X            Y  \
0        18906        STRAITS BOULEVARD  29813.66349  28697.52076   
1        18907        STRAITS BOULEVARD  30041.83890  28602.98724   
2        18910     MARINA GARDENS DRIVE  31966.12079  29115.75337   
3        18916        CENTRAL BOULEVARD  30033.60446  29133.10468   
4        18925              PARK STREET  30367.92624  28764.38143   
...        ...                      ...          ...          ...   
141843  886129  UPPER CHANGI ROAD NORTH  42827.06921  37158.18435   
141844  887327              KRANJI LOOP  19690.28100  46149.11839   
141845  887328       PASIR PANJANG ROAD  23828.90274  28848.55342   
141846  918104       AIRPORT CARGO ROAD  46180.46821  39701.53429   
141847  918146        AIRPORT BOULEVARD  45333.00250  37451.71481   

       postal_code_merge_key  
0                     018906  
1                     018907  
2                     018910  
3                     018916  
4                     018925  
...                      ...  
141843                886129  
141844                887327  
141845                887328  
141846                918104  
141847                918146  

[141848 rows x 10 columns]

## Check for completeness with other two data sources

open_data_postal_code

In [14]:
def compare_postal_codes(
    df1, df2, df1_alias, df2_alias, postal_code_col="postal_code_merge_key"
):
    # Get unique postal codes from each DataFrame
    df1_postal_codes = set(df1[postal_code_col].unique())
    df2_postal_codes = set(df2[postal_code_col].unique())

    # Count the number of unique postal codes in each DataFrame
    df1_postal_count = len(df1_postal_codes)
    df2_postal_count = len(df2_postal_codes)

    # Calculate the total number of unique postal codes
    total_postal_count = len(df1_postal_codes.union(df2_postal_codes))

    # Calculate the percentage of postal codes in each DataFrame
    df1_percentage = (df1_postal_count / total_postal_count) * 100
    df2_percentage = (df2_postal_count / total_postal_count) * 100

    # Print the results
    print(f"{df1_alias} has {df1_postal_count} postal codes ({df1_percentage:.2f}%)")
    print(f"{df2_alias} has {df2_postal_count} postal codes ({df2_percentage:.2f}%)")

In [15]:
df1 = onemap_merge
df1_alias = "onemap"
df2 = postal_merge
df2_alas = "Opendata"
compare_postal_codes(df1, df2, df1_alias, df2_alas)

onemap has 121516 postal codes (89.94%)
Opendata has 134406 postal codes (99.48%)


In [16]:
df1 = onemap_merge
df1_alias = "onemap"
df2 = postcode_address_merge
df2_alas = "postcode_address"
compare_postal_codes(df1, df2, df1_alias, df2_alas)

onemap has 121516 postal codes (89.99%)
postcode_address has 124282 postal codes (92.04%)


In [17]:
df1 = postal_merge
df1_alias = "Opendata"
df2 = postcode_address_merge
df2_alas = "postcode_address"
compare_postal_codes(df1, df2, df1_alias, df2_alas)

Opendata has 134406 postal codes (100.00%)
postcode_address has 124282 postal codes (92.47%)


From the above Opendata looks to be the most complete, but it doesn't have much ito of address info. Second is OneMap, though it's scrape is outdated, leaving room for scraping more recent data if we wish.

## Postal comparison

In [18]:
postal_codes_comp = postal_codes_merge.merge(
    onemap_merge, on="postal_code_merge_key", how="left"
)
postal_codes_comp_nan = postal_codes_comp.loc[postal_codes_comp["LATITUDE"].isna()]
postal_codes_comp_nan

country_code postal_code              place_name admin_name1  \
136              SG      828694      Punggol Point Road        None   
409              SG      519320     Pasir Ris Farmway 2        None   
410              SG      519322     Pasir Ris Farmway 2        None   
1076             SG      778792       Lorong Lada Hitam        None   
1349             SG      588175  Upper Bukit Timah Road        None   
...             ...         ...                     ...         ...   
140809           SG      349148           Kallang Way 2        None   
140810           SG      349187           Kallang Way 1        None   
140938           SG      387577             Sims Avenue        None   
140939           SG      387581             Sims Avenue        None   
141349           SG      118576      Pasir Panjang Road        None   

       admin_code1 admin_name2 admin_code2 admin_name3 admin_code3  latitude  \
136           None        None        None        None        None    1.4209   
409           None        None        None        None        None    1.3852   
410           None        None        None        None        None    1.3840   
1076          None        None        None        None        None    1.4162   
1349          None        None        None        None        None    1.3411   
...            ...         ...         ...         ...         ...       ...   
140809        None        None        None        None        None    1.3219   
140810        None        None        None        None        None    1.3214   
140938        None        None        None        None        None    1.3165   
140939        None        None        None        None        None    1.3165   
141349        None        None        None        None        None    1.2790   

        ...  postal_code_merge_key  ADDRESS BLK_NO BUILDING LATITUDE  \
136     ...                 828694      NaN    NaN      NaN      NaN   
409     ...                 519320      NaN    NaN      NaN      NaN   
410     ...                 519322      NaN    NaN      NaN      NaN   
1076    ...                 778792      NaN    NaN      NaN      NaN   
1349    ...                 588175      NaN    NaN      NaN      NaN   
...     ...                    ...      ...    ...      ...      ...   
140809  ...                 349148      NaN    NaN      NaN      NaN   
140810  ...                 349187      NaN    NaN      NaN      NaN   
140938  ...                 387577      NaN    NaN      NaN      NaN   
140939  ...                 387581      NaN    NaN      NaN      NaN   
141349  ...                 118576      NaN    NaN      NaN      NaN   

       LONGITUDE POSTAL  ROAD_NAME   X   Y  
136          NaN    NaN        NaN NaN NaN  
409          NaN    NaN        NaN NaN NaN  
410          NaN    NaN        NaN NaN NaN  
1076         NaN    NaN        NaN NaN NaN  
1349         NaN    NaN        NaN NaN NaN  
...          ...    ...        ...  ..  ..  
140809       NaN    NaN        NaN NaN NaN  
140810       NaN    NaN        NaN NaN NaN  
140938       NaN    NaN        NaN NaN NaN  
140939       NaN    NaN        NaN NaN NaN  
141349       NaN    NaN        NaN NaN NaN  

[389 rows x 23 columns]

In [24]:
postal_codes_comp

country_code postal_code       place_name admin_name1 admin_code1  \
0                SG      806485     Sunrise Lane        None        None   
1                SG      806485     Sunrise Lane        None        None   
2                SG      806526    Sunrise Drive        None        None   
3                SG      806582    Sunrise Drive        None        None   
4                SG      806605    Sunrise Close        None        None   
...             ...         ...              ...         ...         ...   
141505           SG      138681       Dover Rise        None        None   
141506           SG      138698      Woking Road        None        None   
141507           SG      138811  Whitchurch Road        None        None   
141508           SG      138883    Weyhill Close        None        None   
141509           SG      138941  Westbourne Road        None        None   

       admin_name2 admin_code2 admin_name3 admin_code3  latitude  ...  \
0             None        None        None        None    1.3898  ...   
1             None        None        None        None    1.3898  ...   
2             None        None        None        None    1.3880  ...   
3             None        None        None        None    1.3905  ...   
4             None        None        None        None    1.3875  ...   
...            ...         ...         ...         ...       ...  ...   
141505        None        None        None        None    1.3056  ...   
141506        None        None        None        None    1.2967  ...   
141507        None        None        None        None    1.2938  ...   
141508        None        None        None        None    1.2986  ...   
141509        None        None        None        None    1.2943  ...   

        postal_code_merge_key  \
0                      806485   
1                      806485   
2                      806526   
3                      806582   
4                      806605   
...                       ...   
141505                 138681   
141506                 138698   
141507                 138811   
141508                 138883   
141509                 138941   

                                                  ADDRESS BLK_NO  \
0                       24 SUNRISE LANE SINGAPORE 806485      24   
1          24 SUNRISE LANE FLORIDA PARK SINGAPORE 806485      24   
2        23 SUNRISE DRIVE SUNRISE VILLA SINGAPORE 806526      23   
3         81 SUNRISE DRIVE FLORIDA PARK SINGAPORE 806582      81   
4         8 SUNRISE CLOSE SUNRISE VILLA SINGAPORE 806605       8   
...                                                   ...    ...   
141505     34 DOVER RISE DOVER PARKVIEW SINGAPORE 138681      34   
141506            20 WOKING ROAD WESSEX SINGAPORE 138698      20   
141507         3 WHITCHURCH ROAD WESSEX SINGAPORE 138811       3   
141508           3 WEYHILL CLOSE WESSEX SINGAPORE 138883       3   
141509         1 WESTBOURNE ROAD WESSEX SINGAPORE 138941       1   

              BUILDING  LATITUDE   LONGITUDE    POSTAL        ROAD_NAME  \
0                  NIL  1.389751  103.854623  806485.0     SUNRISE LANE   
1         FLORIDA PARK  1.389839  103.854636  806485.0     SUNRISE LANE   
2        SUNRISE VILLA  1.388029  103.857166  806526.0    SUNRISE DRIVE   
3         FLORIDA PARK  1.390531  103.854870  806582.0    SUNRISE DRIVE   
4        SUNRISE VILLA  1.387482  103.857554  806605.0    SUNRISE CLOSE   
...                ...       ...         ...       ...              ...   
141505  DOVER PARKVIEW  1.305622  103.783822  138681.0       DOVER RISE   
141506          WESSEX  1.296666  103.795198  138698.0      WOKING ROAD   
141507          WESSEX  1.293838  103.798243  138811.0  WHITCHURCH ROAD   
141508          WESSEX  1.298591  103.793000  138883.0    WEYHILL CLOSE   
141509          WESSEX  1.294349  103.796537  138941.0  WESTBOURNE ROAD   

                  X            Y  
0       30370.94726  41297.18305  
1       30372.38505  41306.88482

It's only 389, so it should be fine for now. And a lot of them are actually in the secondary dataset, at least for the purpose of addresses.

In [19]:
postal_codes_comp_nan["postal_code_merge_key"].isin(
    postcode_address_merge["postal_code_merge_key"].values
).sum()

np.int64(314)

In [34]:
postcode_address_merge_supp = postcode_address_merge.loc[
    postcode_address_merge["postal_code_merge_key"].isin(
        postal_codes_comp_nan["postal_code_merge_key"]
    )
]
postcode_address_merge_supp = postcode_address_merge_supp.assign(
    address_info=postcode_address_merge_supp["address"]
    .str.split(",")
    .apply(lambda x: len(x))
)[["address", "address_info"]]
postcode_address_merge_supp

address  address_info
314     Kallang Way 1, 135, Singapore, Kallang, Woodle...             8
535     Fidvi Court, Joo Chiat Terrace, 144, Singapore...             8
699     Sims Drive, 54, Singapore, Sims, Kallang, Alju...             8
1713    Tampines Industrial Street 62, 4, Singapore, T...             7
2083    Goh & Goh Building, Upper Bukit Timah Road, 12...             8
...                                                   ...           ...
121892  Townhouse Condominium, How Sun Road, 45, Singa...            10
122457  Townhouse Condominium, How Sun Road, 61, Singa...            10
122511  Normanton Park, Normanton Park, 2, Singapore, ...             9
122764  The Aquarius, Science Park Road, 21, Singapore...             9
123042  Jalan Besar, 257-2B, Singapore, Besar, Lavende...             7

[314 rows x 2 columns]

And for most, we can get the postcode info from the above source...

## Geocode via OneMap

But, given it's rather dubious source, we can geocode it directly instead.

In [35]:
from pathlib import Path

from kedro.config import OmegaConfigLoader
from kedro.framework.project import settings

# Substitute <project_root> with the [root folder for your project](https://docs.kedro.org/en/stable/tutorial/spaceflights_tutorial.html#terminology)
conf_path = str(Path(context.project_path) / settings.CONF_SOURCE)
conf_loader = OmegaConfigLoader(conf_source=conf_path)
credentials = conf_loader["credentials"]
one_map_access_token = credentials["one_map_access_token"]

In [36]:
import requests

url = "https://www.onemap.gov.sg/api/common/elastic/search?searchVal=349182&returnGeom=Y&getAddrDetails=Y&pageNum=1"
headers = {"Authorization": one_map_access_token}
response = requests.get(url, headers=headers)
print(response.json())

{'found': 0, 'totalNumPages': 0, 'pageNum': 1, 'results': []}


This doesn't return anything, and we are only really after address info...

In [37]:
results = []
codes = postal_codes_comp_nan["postal_code_merge_key"].unique()
for i, code in enumerate(postal_codes_comp_nan["postal_code_merge_key"].unique()):
    print(f"{code}: {i + 1} of {len(codes)}")
    url = f"https://www.onemap.gov.sg/api/common/elastic/search?searchVal={code}&returnGeom=Y&getAddrDetails=Y&pageNum=1"
    response = requests.get(url, headers=headers)
    results.append(response.json())

828694: 1 of 389
519320: 2 of 389
519322: 3 of 389
778792: 4 of 389
588175: 5 of 389
598665: 6 of 389
637372: 7 of 389
737738: 8 of 389
349063: 9 of 389
399761: 10 of 389
098330: 11 of 389
098844: 12 of 389
098850: 13 of 389
118496: 14 of 389
139246: 15 of 389
238436: 16 of 389
259712: 17 of 389
058924: 18 of 389
089539: 19 of 389
118574: 20 of 389
423759: 21 of 389
426096: 22 of 389
209194: 23 of 389
347769: 24 of 389
349182: 25 of 389
369584: 26 of 389
390026: 27 of 389
275806: 28 of 389
275808: 29 of 389
275814: 30 of 389
275820: 31 of 389
419805: 32 of 389
538511: 33 of 389
538512: 34 of 389
538516: 35 of 389
739070: 36 of 389
329756: 37 of 389
369581: 38 of 389
069934: 39 of 389
118575: 40 of 389
275797: 41 of 389
275811: 42 of 389
275813: 43 of 389
275819: 44 of 389
498741: 45 of 389
506976: 46 of 389
528815: 47 of 389
538502: 48 of 389
207390: 49 of 389
238433: 50 of 389
239927: 51 of 389
349102: 52 of 389
349108: 53 of 389
349183: 54 of 389
390030: 55 of 389
275801: 56 of 389
2

So only 5 of these are now in OpenMap.

In [41]:
len([r for r in results if r["found"]])

5

## Testing accuracy of postal code matching compared to onemap

In [47]:
postal_codes_comp.columns


Index(['country_code', 'postal_code', 'place_name', 'admin_name1',
       'admin_code1', 'admin_name2', 'admin_code2', 'admin_name3',
       'admin_code3', 'latitude', 'longitude', 'accuracy', 'coordinates',
       'postal_code_merge_key', 'ADDRESS', 'BLK_NO', 'BUILDING', 'LATITUDE',
       'LONGITUDE', 'POSTAL', 'ROAD_NAME', 'X', 'Y'],
      dtype='object')

postal_codes_comp.columns

In [ ]:
import geopandas as gpd

geo1 = gpd.GeoDataFrame(
    postal_codes_comp,
    geometry=gpd.points_from_xy(
        postal_codes_comp["LONGITUDE"], postal_codes_comp["LATITUDE"]
    ),
    crs="EPSG:4326",
).to_crs("EPSG:3857")
geo2 = gpd.GeoDataFrame(
    postal_codes_comp,
    geometry=gpd.points_from_xy(
        postal_codes_comp["longitude"], postal_codes_comp["latitude"]
    ),
    crs="EPSG:4326",
).to_crs("EPSG:3857")

In [63]:
postal_codes_comp = postal_codes_comp.assign(geo_dist=geo1.distance(geo2))
postal_codes_comp.sort_values("geo_dist", ascending=False)[
    [
        "place_name",
        "ADDRESS",
        "geo_dist",
        "postal_code_merge_key",
        "latitude",
        "longitude",
        "LATITUDE",
        "LONGITUDE",
    ]
].head(25)

place_name  \
40716                  Tampines Street 32   
40718                  Tampines Street 32   
62985                  Tampines Central 1   
40714                  Tampines Street 32   
40717                  Tampines Street 32   
40715                  Tampines Street 32   
86031   Pasir Ris Coast Industrial Park 6   
111078                         Cluny Road   
1629                     Nanyang Crescent   
87277                       Merlimau Road   
125128                Kent Ridge Crescent   
22558                    Jurong Port Road   
22551                    Jurong Port Road   
20518                       Pulau Sebarok   
22563                    Jurong Port Road   
22567                    Jurong Port Road   
103469                 Pulau Tekong Besar   
107830                      Harbour Drive   
26258               Lower Kent Ridge Road   
103468                 Pulau Tekong Besar   
54634                Tuas South Boulevard   
66133                     Chestnut Avenue   
54633                Tuas South Boulevard   
22555                    Jurong Port Road   
22566                    Jurong Port Road   

                                                  ADDRESS     geo_dist  \
40716    20 TAMPINES CENTRAL 1 CITIBANK TAMPINES SMRT ...  1664.813143   
40718    20 TAMPINES CENTRAL 1 DBS TAMPINES MRT STATIO...  1659.693636   
62985    9 TAMPINES STREET 32 DBS TAMPINES MART SINGAP...  1653.742789   
40714    20 TAMPINES CENTRAL 1 TAMPINES MRT STATION SI...  1645.559939   
40717    20 TAMPINES CENTRAL 1 OCBC TAMPINES MRT STATI...  1631.863241   
40715    20 TAMPINES CENTRAL 1 DBS TAMPINES MRT STATIO...  1620.561443   
86031    101 PASIR RIS COAST INDUSTRIAL PARK 6 SERANGO...  1210.369926   
111078                     1 CLUNY ROAD SINGAPORE 259569   1086.883900   
1629     50 NANYANG CRESCENT NANYANG TECHNOLOGICAL UNI...   965.675124   
87277                   1 MERLIMAU ROAD SINGAPORE 628260    952.921798   
125128   10 KENT RIDGE CRESCENT NATIONAL UNIVERSITY OF...   883.236490   
22558    37 JURONG PORT ROAD J17 (JURONG PORT) SINGAPO...   872.759739   
22551    37 JURONG PORT ROAD J4 (JURONG PORT) SINGAPOR...   829.426888   
20518                   1 PULAU SEBAROK SINGAPORE 118326    819.230796   
22563    37 JURONG PORT ROAD J16 (JURONG PORT) SINGAPO...   818.696235   
22567    37 JURONG PORT ROAD J3 (JURONG PORT) SINGAPOR...   782.387970   
103469   21 PULAU TEKONG BESAR LADANG CAMP SINGAPORE 5...   772.798012   
107830   33 HARBOUR DRIVE PASIR PANJANG TERMINAL BUILD...   770.103835   
26258    21 LOWER KENT RIDGE ROAD NATIONAL UNIVERSITY ...   769.506773   
103468   21 PULAU TEKONG BESAR ROCKY HILL CAMP SINGAPO...   745.524359   
54634    80 TUAS SOUTH BOULEVARD SEMBMARINE INTEGRATED...   725.859629   
66133               201 CHESTNUT AVENUE SINGAPORE 679525    701.170597   
54633    80 TUAS SOUTH BOULEVARD DBS SEMBCORP MARINE S...   694.098903   
22555    37 JURONG PORT ROAD J5 (JURONG PORT) SINGAPOR...   665.989624   
22566    37 JURONG PORT ROAD J2 (JURONG PORT) SINGAPOR...   660.158518   

       postal_code_merge_key  latitude  longitude  LATITUDE   LONGITUDE  
40716                 529538    1.3542   103.9599  1.353547  103.944959  
40718                 529538    1.3542   103.9599  1.353571  103.945004  
62985                 529538    1.3534   103.9451  1.354167  103.959936  
40714                 529538    1.3542   103.9599  1.353302  103.945145  
40717                 529538    1.3542   103.9599  1.353101  103.945282  
40715                 529538    1.3542   103.9599  1.353195  103.945377  
86031                 518200    1.4084   103.9221  1.400582  103.929654  
111078                259569    1.3126   103.8162  1.322157  103.818186  
1629                  637598    1.3470   103.6855  1.355540  103.687010  
87277                 628260    1.2816   103.7140  1.290125  103.713248  
125128                119260    1.2969   103.7726  1.294521  103.780169  
22558                 619110    1.3065   103.72

Based on the above, we can see that the postal codes from the OneMap API and the OpenData postal codes are not a perfect match. The postal codes from the OneMap API are more accurate as they are based on the actual latitude and longitude coordinates of the postal codes. The OpenData postal codes are based on the postal code itself, which may not always be accurate. We can use the OneMap API to get the latitude and longitude coordinates of the postal codes and use that information to update the OpenData postal codes. It could be that there is geocoding for the address plus postal codes with the first data source. Whereas onemap should be more accurate.

## Final pipeline

Use OneMap, for NaNs, combine Opendata with Postcode base.

In [66]:
onemap  # assign SOURCE = "OneMap"

ADDRESS   BLK_NO  \
0        1 STRAITS BOULEVARD SINGAPORE CHINESE CULTURA...        1   
1        11A STRAITS BOULEVARD TEMPORARY SITE OFFICE S...      11A   
2               5A MARINA GARDENS DRIVE SINGAPORE 018910        5A   
3        2 CENTRAL BOULEVARD CENTRAL BOULEVARD TOWERS ...        2   
4        21 PARK STREET MARINA BAY MRT STATION SINGAPO...       21   
...                                                   ...      ...   
141843       1A UPPER CHANGI ROAD NORTH SINGAPORE 886129        1A   
141844                 100A KRANJI LOOP SINGAPORE 887327      100A   
141845             A PASIR PANJANG ROAD SINGAPORE 887328         A   
141846   GATE C7 AIRPORT CARGO ROAD CHANGI ANIMAL AND ...  GATE C7   
141847   AIRPORT BOULEVARD CHANGI AIRPORT TERMINAL 2 B...      NaN   

                                         BUILDING  LATITUDE   LONGITUDE  \
0               SINGAPORE CHINESE CULTURAL CENTRE  1.275805  103.849615   
1                           TEMPORARY SITE OFFICE  1.274950  103.851665   
2                                             NIL  1.279587  103.868956   
3                        CENTRAL BOULEVARD TOWERS  1.279744  103.851591   
4                   MARINA BAY MRT STATION (NS27)  1.276409  103.854595   
...                                           ...       ...         ...   
141843                                        NIL  1.352316  103.966549   
141844                                        NIL  1.433629  103.758648   
141845                                        NIL  1.277170  103.795840   
141846  CHANGI ANIMAL AND PLANT QUARANTINE CENTRE  1.375315  103.996683   
141847         CHANGI AIRPORT TERMINAL 2 BUILDING  1.354969  103.989067   

        POSTAL                ROAD_NAME            X            Y  
0        18906        STRAITS BOULEVARD  29813.66349  28697.52076  
1        18907        STRAITS BOULEVARD  30041.83890  28602.98724  
2        18910     MARINA GARDENS DRIVE  31966.12079  29115.75337  
3        18916        CENTRAL BOULEVARD  30033.60446  29133.10468  
4        18925              PARK STREET  30367.92624  28764.38143  
...        ...                      ...          ...          ...  
141843  886129  UPPER CHANGI ROAD NORTH  42827.06921  37158.18435  
141844  887327              KRANJI LOOP  19690.28100  46149.11839  
141845  887328       PASIR PANJANG ROAD  23828.90274  28848.55342  
141846  918104       AIRPORT CARGO ROAD  46180.46821  39701.53429  
141847  918146        AIRPORT BOULEVARD  45333.00250  37451.71481  

[141848 rows x 9 columns]

In [153]:
import pandas as pd

def get_postal_codes(postal_codes):
    return postal_codes.astype("string").str.zfill(6)


def format_onemap(df):
    return df.assign(
        POSTAL=get_postal_codes(df["POSTAL"]),
        SOURCE="OneMap",
        URL="https://www.onemap.gov.sg/apidocs/search",
        URL_2="https://github.com/xuancong84/singapore-address-heatmap/blob/master/database.csv.gz",
    ).drop(columns=["X", "Y"])


def format_opendata(df):
    df = df.loc[df["country_code"] == "SG"]
    df = df.assign(postal_code=get_postal_codes(df["postal_code"]))
    df = df.assign(
        ADDRESS=(df["place_name"] + " singapore " + df["postal_code"])
        .str.replace(",", "")
        .str.upper(),
        ROAD_NAME=df["place_name"].str.replace(",", "").str.upper(),
    )
    return df.rename(
        columns={
            "postal_code": "POSTAL",
            "latitude": "LATITUDE",
            "longitude": "LONGITUDE",
        },
        inplace=False,
    )[["ADDRESS", "ROAD_NAME", "LATITUDE", "LONGITUDE", "POSTAL"]].assign(
        SOURCE="opendatasoft",
        URL="https://public.opendatasoft.com/explore/dataset/geonames-postal-code/table/?refine.country_code=SG",
    )


def format_postcodebase(df):
    df = df.assign(postcode=get_postal_codes(df["postcode"]))
    df = df.assign(
        ADDRESS=(
            df["address"].str.replace(r" is located in Singapore.*$", "", regex=True)
            + " singapore "
            + df["postcode"]
        )
        .str.replace(",", "")
        .str.upper()
    )
    return df.rename(
        columns={
            "postcode": "POSTAL",
        },
        inplace=False,
    )[["ADDRESS", "POSTAL"]].assign(
        SOURCE="Postcodebase",
        URL="https://sgp.postcodebase.com/all",
    )


def enrich_open_postcode(open_postcode, postcode_base):
    return open_postcode.drop(columns=["ADDRESS"]).merge(
        postcode_base,
        left_on="POSTAL",
        right_on="POSTAL",
        how="left",
        validate="m:1",
        suffixes=("", "_2"),
    )


def extend_onemap(onemap_postal_codes, open_postcode_enriched):
    additional_postal_codes = open_postcode_enriched.loc[~open_postcode_enriched["POSTAL"].isin(onemap_postal_codes["POSTAL"].values)].reset_index(drop=True)
    onemap_postal_codes = pd.concat([onemap_postal_codes, additional_postal_codes], ignore_index=True, axis=0)
    return onemap_postal_codes


def postcode_full_type_conversion(df):
    rename_columns = {
            "POSTAL": "string",
            "ADDRESS": "string",
            "BLK_NO": "string",
            "BUILDING": "string",
            "ROAD_NAME": "string",
            "LATITUDE": "float",
            "LONGITUDE": "float",
            "SOURCE": "string",
            "URL": "string",
            "SOURCE_2": "string",
            "URL_2": "string",
        }
    return df.astype(
        rename_columns
    )[list(rename_columns.keys())]

onemap = catalog.load("one_map_scrape")
open_postcode = catalog.load("open_data_postal_code")
postcode_base = catalog.load("sg_postcode_based_via_getdata")

onemap_postal_codes = format_onemap(onemap)
open_postcode = format_opendata(open_postcode)
postcode_base = format_postcodebase(postcode_base)

open_postcode_enriched = enrich_open_postcode(open_postcode, postcode_base)
onemap_postal_codes_extended = extend_onemap(onemap_postal_codes, open_postcode_enriched).pipe(postcode_full_type_conversion)
onemap_postal_codes_extended

[01/22/25 22:05:52] INFO     Loading data from one_map_scrape (CSVDataset)...                   ]8;id=657801;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=17365;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py#389\389]8;;\

                    INFO     Loading data from open_data_postal_code (ParquetDataset)...        ]8;id=549841;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=43219;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py#389\389]8;;\

[01/22/25 22:05:53] INFO     Loading data from sg_postcode_based_via_getdata (CSVDataset)...    ]8;id=707890;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=502980;file:///Users/ejwillemse/dev/projects/second_order_ai/singapore-postcode-geocoding/.venv/lib/python3.12/site-packages/kedro/io/data_catalog.py#389\389]8;;\

POSTAL                                            ADDRESS BLK_NO  \
0       018906   1 STRAITS BOULEVARD SINGAPORE CHINESE CULTURA...      1   
1       018907   11A STRAITS BOULEVARD TEMPORARY SITE OFFICE S...    11A   
2       018910          5A MARINA GARDENS DRIVE SINGAPORE 018910      5A   
3       018916   2 CENTRAL BOULEVARD CENTRAL BOULEVARD TOWERS ...      2   
4       018925   21 PARK STREET MARINA BAY MRT STATION SINGAPO...     21   
...        ...                                                ...    ...   
142232  349148  KALLANG WAY 2 125 SINGAPORE KALLANG WOODLEIGH ...   <NA>   
142233  349187  KALLANG WAY 1 145 SINGAPORE KALLANG WOODLEIGH ...   <NA>   
142234  387577  SIMS AVENUE 517 SINGAPORE SIMS KALLANG ALJUNIE...   <NA>   
142235  387581  SIMS AVENUE 521 SINGAPORE SIMS KALLANG ALJUNIE...   <NA>   
142236  118576  HARBOUR VIEW GARDENS PASIR PANJANG ROAD 215 SI...   <NA>   

                                 BUILDING             ROAD_NAME  LATITUDE  \
0       SINGAPORE CHINESE CULTURAL CENTRE     STRAITS BOULEVARD  1.275805   
1                   TEMPORARY SITE OFFICE     STRAITS BOULEVARD  1.274950   
2                                     NIL  MARINA GARDENS DRIVE  1.279587   
3                CENTRAL BOULEVARD TOWERS     CENTRAL BOULEVARD  1.279744   
4           MARINA BAY MRT STATION (NS27)           PARK STREET  1.276409   
...                                   ...                   ...       ...   
142232                               <NA>         KALLANG WAY 2  1.321900   
142233                               <NA>         KALLANG WAY 1  1.321400   
142234                               <NA>           SIMS AVENUE  1.316500   
142235                               <NA>           SIMS AVENUE  1.316500   
142236                               <NA>    PASIR PANJANG ROAD  1.279000   

         LONGITUDE        SOURCE  \
0       103.849615        OneMap   
1       103.851665        OneMap   
2       103.868956        OneMap   
3       103.851591        OneMap   
4       103.854595        OneMap   
...            ...           ...   
142232  103.877200  opendatasoft   
142233  103.876900  opendatasoft   
142234  103.889100  opendatasoft   
142235  103.889200  opendatasoft   
142236  103.785200  opendatasoft   

                                                      URL      SOURCE_2  \
0                https://www.onemap.gov.sg/apidocs/search          <NA>   
1                https://www.onemap.gov.sg/apidocs/search          <NA>   
2                https://www.onemap.gov.sg/apidocs/search          <NA>   
3                https://www.onemap.gov.sg/apidocs/search          <NA>   
4                https://www.onemap.gov.sg/apidocs/search          <NA>   
...                                                   ...           ...   
142232  https://public.opendatasoft.com/explore/datase...  Postcodebase   
142233  https://public.opendatasoft.com/explore/datase...  Postcodebase   
142234  https://public.opendatasoft.com/explore/datase...  Postcodebase   
142235  https://public.opendatasoft.com/explore/datase...  Postcodebase   
142236  https://public.opendatasoft.com/explore/datase...  Postcodebase   

                                                    URL_2  
0       https://github.com/xuancong84/singapore-addres...  
1       https://github.com/xuancong84/singapore-addres...  
2       https://github.com/xuancong84/singapore-addres...  
3       https://github.com/xuancong84/singapore-addres...  
4       https://github.com/xuancong84/singapore-addres...  
...                                                   ...  
142232                   https://sgp.postcodebase.com/all  
142233                   https://sgp.postcodebase.com/all  
142234                   https://sgp.postcodebase.com/all  
142235                   https://sgp.postcodebase.com/all  
142236                   https://sgp.postcodebase.com/all  

[142237 rows x 11 columns]